google/cloud/security/scanner/audit/forwarding_rule_rules_engine.py

```diff
@@ -0,0 +1,253 @@
+                target=forwarding_rule.target,
+                port_range=forwarding_rule.port_range,
+                port=forwarding_rule.ports,
+                ip_protocol=forwarding_rule.ip_protocol,
+                ip_address=forwarding_rule.ip_address,
+                resource_id=forwarding_rule.resource_id,
+                rule_index=len(resource_rules),
+                resource_type=ResourceType.FORWARDING_RULE)
+
+    # Rule violation.
+    # rule_name: string
+    # rule_index: int
+    # violation_type: FORWARDING_RULE_VIOLATION
+    # target: string
+    # load_balancing_scheme: string
+    # port_range: string
+    # port: string
+    # ip_protocol: string
+    # ip_address: string
+    RuleViolation = namedtuple('RuleViolation',
+                               [
+                                   'violation_type', 'target', 'rule_index',
+                                   'load_balancing_scheme', 'port_range',
+                                   'resource_type', 'port', 'ip_protocol',
+                                   'ip_address', 'resource_id'])
+
+    def add_rules(self, rules):
+        """Add rules to the rule book."""
+        if self.rule_book is not None:
+            self.rule_book.add_rules(rules)
+
+
+class ForwardingRuleRulesBook(bre.BaseRuleBook):
+    """The RuleBook for bucket acls resources."""
+
+    def __init__(self, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs: rule definitons
+        """
+        super(ForwardingRuleRulesBook, self).__init__()
+        self.resource_rules_map = {}
+        if not rule_defs:
+            self.rule_defs = {}
+        else:
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book"""
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def: A dictionary containing rule definition properties.
+            rule_index: The index of the rule from the rule definitions.
+            Assigned automatically when the rule book is built.
+
+        Raises:
+
+        """
+
+        target = rule_def.get('target')
+        mode = rule_def.get('mode')
+        load_balancing_scheme = rule_def.get('load_balancing_scheme')
+        port_range = rule_def.get('port_range')
+        port = rule_def.get('port')
+        ip_address = rule_def.get('ip_address')
+        ip_protocol = rule_def.get('ip_protocol')
+        if (target is None) or (mode is None) or \
+           (load_balancing_scheme is None) or \
+           (ip_address is None) or (ip_protocol is None):
+            raise audit_errors.InvalidRulesSchemaError(
+                'Faulty rule {}'.format(rule_def.get('name')))
+        rule_def_resource = {'target': target,
```

Comment:
Food for thought, for readability purpose isn't it better to use the syntax rule_def_resource['target'] = target?